lib/ble_conn_state/ble_conn_state.c

```diff
@@ -0,0 +1,424 @@
+}
+
+enum ble_conn_state_status ble_conn_state_status(uint16_t conn_handle)
+{
+	int idx = nrf_sdh_ble_idx_get(conn_handle);
+	bool connected;
+	enum ble_conn_state_status conn_status = BLE_CONN_STATUS_INVALID;
+
+	if (!ble_conn_state_valid_idx(idx)) {
+		return BLE_CONN_STATUS_INVALID;
+	}
+
+	connected = atomic_test_bit(&bcs.flags.connected_flags, idx);
+	conn_status = connected ? BLE_CONN_STATUS_CONNECTED : BLE_CONN_STATUS_DISCONNECTED;
+
+	return conn_status;
+}
+
+bool ble_conn_state_encrypted(uint16_t conn_handle)
+{
+	int idx = nrf_sdh_ble_idx_get(conn_handle);
+
+	if (!ble_conn_state_valid_idx(idx)) {
+		return false;
+	}
+
+	return atomic_test_bit(&bcs.flags.encrypted_flags, idx);
+}
+
+bool ble_conn_state_mitm_protected(uint16_t conn_handle)
+{
+	int idx = nrf_sdh_ble_idx_get(conn_handle);
+
+	if (!ble_conn_state_valid_idx(idx)) {
+		return false;
+	}
+
+	return atomic_test_bit(&bcs.flags.mitm_protected_flags, idx);
+}
+
+bool ble_conn_state_lesc(uint16_t conn_handle)
+{
+	int idx = nrf_sdh_ble_idx_get(conn_handle);
+
+	if (!ble_conn_state_valid_idx(idx)) {
+		return false;
+	}
+
+	return atomic_test_bit(&bcs.flags.lesc_flags, idx);
+}
+
+uint32_t ble_conn_state_conn_count(void)
+{
+	return active_flag_count(bcs.flags.connected_flags);
+}
+
+uint32_t ble_conn_state_central_conn_count(void)
+{
+	atomic_t central_conn_flags;
+
+	central_conn_flags = bcs.flags.central_flags;
+	(void)atomic_and(&central_conn_flags, bcs.flags.connected_flags);
+
+	return active_flag_count(central_conn_flags);
+}
+
+uint32_t ble_conn_state_peripheral_conn_count(void)
+{
+	atomic_t peripheral_conn_flags;
+
+	peripheral_conn_flags = ~bcs.flags.central_flags;
+	(void)atomic_and(&peripheral_conn_flags, bcs.flags.connected_flags);
+
+	return active_flag_count(peripheral_conn_flags);
+}
+
+struct ble_conn_state_conn_handle_list ble_conn_state_conn_handles(void)
+{
+	return conn_handle_list_get(bcs.flags.valid_flags);
+}
+
+struct ble_conn_state_conn_handle_list ble_conn_state_central_handles(void)
+{
+	atomic_t central_conn_flags;
+
+	central_conn_flags = bcs.flags.central_flags;
+	(void)atomic_and(&central_conn_flags, bcs.flags.connected_flags);
+
+	return conn_handle_list_get(central_conn_flags);
+}
+
+struct ble_conn_state_conn_handle_list ble_conn_state_periph_handles(void)
+{
+	atomic_t peripheral_conn_flags;
+
+	peripheral_conn_flags = ~bcs.flags.central_flags;
+	(void)atomic_and(&peripheral_conn_flags, bcs.flags.connected_flags);
+
+	return conn_handle_list_get(peripheral_conn_flags);
+}
+
+uint16_t ble_conn_state_conn_idx(uint16_t conn_handle)
+{
+	int idx = nrf_sdh_ble_idx_get(conn_handle);
+
+	if (!ble_conn_state_valid_idx(idx)) {
+		return BLE_CONN_STATE_MAX_CONNECTIONS;
+	}
+
+
+	return idx;
+}
+
+int ble_conn_state_user_flag_acquire(void)
+{
+	for (int i = 0; i < CONFIG_BLE_CONN_STATE_USER_FLAG_COUNT; i++) {
+		if (atomic_test_and_set_bit(&bcs.acquired_flags, i) == false) {
+			/* Bit was not set before */
+			return i;
+		}
+	}
+
+	return -ENOSPC;
```

Comment:
This is going against the recent discussions we've had about error codes direction for BLE related components. Is there some other way to handle this? Since Peer Manager depends on this component, I would be careful with changing return values or all the changes will have to be propagated to the Peer Manager and will have to be debugged at runtime due to possible subtle bugs (since `uint32_t` vs `int` compiles ok).

Reply:
Discussed online and agreed. Replaced with BLE_CONN_STATE_USER_FLAG_INVALID.